### PR TITLE
Fix DefaultRestChannel Corrupting Shared Buffers on Serialialization Issues

### PR DESCRIPTION
**Pull request by original-brownbear**
We must not reset the shared buffer after it has been used (can happen in error handling in `RestController#sendResponse`).
There is never a good reason to reset a pooled bytes output either and the behavior isn't clearly defined so this commit
disables the operation as it had unintended side effects.

NOTE to reviewers:

I'm aware that the existing code is extremely hard to follow in terms of the buffer lifecycle and this change doesn't necessarily help things. Then again, the bug fixed here is potentially very impactful for users and it would be good to backport to 7.13 so I tried to go for the least number of changes/refactoring possible instead of refactoring things to make the buffer handling more easy to follow which we can do in a follow-up instead.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -1550,26 +1551,20 @@ static void verifyChecksum(BufferedChecksumStreamInput in) throws IOException {
      * use {@link #readOperations(StreamInput, String)} to read it back.
      */
     public static void writeOperations(StreamOutput outStream, List<Operation> toWrite) throws IOException {
-        final ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(BigArrays.NON_RECYCLING_INSTANCE);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java:1553]: This was the one place where we reset a releasable bytes output but it's obviously a bogus spot since it's based on non-recycling.

File: server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java [hunk 2]
```diff
@@ -137,25 +141,30 @@ public XContentBuilder newBuilder(@Nullable XContentType requestContentType, @Nu
...
-        if (bytesOut == null) {
-            bytesOut = newBytesOutput();
-        } else {
-            bytesOut.reset();
+        if (bytesOut != null) {
+            // fallback in case of encountering a bug, release the existing buffer if any (to avoid leaking memory) and acquire a new one
+            // to send out an error response
+            assert false : "getting here is always a bug";
+            logger.error("channel handling [{}] reused", request.rawPath());
+            releaseOutputBuffer();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java:153]: I think we need this in practice to prevent bad things from happening in case of serialisation bugs (you don't even have to get into `sendResponse` to acquire the buffer here). I don't want to throw `AssertionError` here and potentially kill the node, throw anything and thus leak memory or other resources associated with the channel -> this was the next best option I could think of.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java [hunk 3]
```diff
@@ -780,7 +780,11 @@ protected ExceptionThrowingChannel(RestRequest request, boolean detailedErrorsEn
 
         @Override
         public void sendResponse(RestResponse response) {
-            throw new IllegalStateException("always throwing an exception for testing");
+            try {
+                throw new IllegalStateException("always throwing an exception for testing");
+            } finally {
+                releaseOutputBuffer();
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:786]: Meh, best short term fix I could think of, normally this would be taken care of by the `DefaultRestChannel` itself, but I suppose it's good that tests trip without this change.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/service/RestClearServiceAccountTokenStoreCacheActionTests.java [hunk 4]
```diff
@@ -50,7 +51,9 @@ public void init() {
         verifyingClient.setExecuteVerifier(((actionType, actionRequest) -> {
             assertThat(actionRequest, instanceOf(ClearSecurityCacheRequest.class));
             requestHolder.set((ClearSecurityCacheRequest) actionRequest);
-            return mock(ClearSecurityCacheResponse.class);
+            final ClearSecurityCacheResponse response = mock(ClearSecurityCacheResponse.class);
+            when(response.getClusterName()).thenReturn(new ClusterName(""));
+            return response;
```
Comment (original-brownbear) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/service/RestClearServiceAccountTokenStoreCacheActionTests.java:56]: This test was pretty bogus (and still is to some degree) since it would always NPE in the `sendResponse` logic without a `ClusterName` returned which now trips an assertion so I fixed the NPE.

File: server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java [hunk 5]
```diff
@@ -90,6 +90,7 @@ public void sendResponse(RestResponse restResponse) {
             if (content instanceof Releasable) {
                 toClose.add((Releasable) content);
             }
+            toClose.add(this::releaseOutputBuffer);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java:93]: This should really be right at the beginning of this method body, if we get here we already have the buffer all set up. If we do it later an e.g. the cors logic NPEs we leak memory.

File: server/src/main/java/org/elasticsearch/common/io/stream/ReleasableBytesStreamOutput.java [hunk 6]
```diff
@@ -37,4 +37,10 @@ public ReleasableBytesStreamOutput(int expectedSize, BigArrays bigArrays) {
     public void close() {
         Releasables.close(bytes);
     }
+
+    @Override
+    public void reset() {
+        // not supported, close and create a new instance instead
+        throw new UnsupportedOperationException("must not reuse a pooled bytes backed stream");
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/io/stream/ReleasableBytesStreamOutput.java:44]: To me it doesn't look like down-sizing a pooled `BigArray` or so works correctly. This is the safest option I could find for the time being.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java:1553]: I wonder how much we really gain from skipping around like this vs just writing two things to the stream. We only seem to use this code when serialising a batch of operations to send over the wire for recoveries anyway so it's already buffered.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:786]:
```suggestion
                // the production implementation in DefaultRestChannel always releases the output buffer, so we must too
                releaseOutputBuffer();
```

Also should we be doing this in `AssertingChannel` a few lines up? Can we add an assertion somewhere that `sendResponse` really does always release the output buffer?

**Review by DaveCTurner — APPROVED** [review]
LGTM, I left a few thoughts too.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:786]: It's hard to assert that in general because the release happens async in the production implementation (via the listener on the Netty/Transport thread that actually ships out the request).

It seems like we don't need to do this in the asserting channel as we don't have any error situations there. Lets add it when we need it I'd say, keeps the assertions stricter IMO

File: server/src/main/java/org/elasticsearch/common/io/stream/ReleasableBytesStreamOutput.java (hunk 6 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/ReleasableBytesStreamOutput.java:44]: ++

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java:1553]: You're right ... that's kind of bogus. I cleaned it up to use two separarte writes.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:787]: ah yeah good point about the async release, nvm then

**Comment by original-brownbear** [comment]
Thanks David!

**Comment by DaveCTurner** [comment]
D'oh I only just spotted the manual serialialization issue in the title :grin:

**Comment by original-brownbear** [comment]
Ouch sorry about that :)
